### DISCUSSION TITLE
WebRTC as the core communication protocol

### DISCUSSION THREAD

**Discussion by WofWca** [rustdesk/rustdesk#2347] [General]
Let me get it out of the way: this issue is _NOT_  just about adaptive media bitrate, congestion control and stuff.

I do not know what protocol suite is used right now (but looks like something custom), but it seems to me like WebRTC would be a perfect suit(e) here.

A quick introduction to those who aren't familiar with WebRTC:
* Peer-to-peer, secure communication.
* Designed to automatically pierce NAT, (internally (no extra code)).
* Highly standardized, cross-platform (works in browsers as well).
* Simple API. Basically almost as simple as making an HTTP request. [An example program in Rust](https://github.com/webrtc-rs/webrtc/blob/56368e9b5424b606ce6e4208a4536ab914096115/examples/examples/data-channels/data-channels.rs#L63-L200) is ~100 lines.
* Can be used to transfer arbitrary data, both reliably/unreliably and in/out of order.
* Built-in media support, with adaptive bitrate and all that jazz.
* Connection can be established without even needing a server, provided both users are at their PCs and are communicating. With a server the process is as usual.
* When true P2P is not possible (due to both peers having a symmetric NAT), an (also standardized and simple to use relay server (called [TURN](https://github.com/webrtc-rs/webrtc/tree/master/turn))) can be used.

Adopting WebRTC should:
* Simplify the current peer-to-peer code with just calls to WebRTC API.
* Simplify the server's code (thanks to TURN).
* Solve (almost?) all the existing (and future) P2P problems.
* Solve the adaptive bitrate video problem, and other problems related to video, in the future. For now the current media data can be transferred as binary data, without touching the WebRTC's media API.

I have not made much research on whether [alternative remote desktop software](https://alternativeto.net/software/rustdesk/) utilizes WebRTC. Perhaps this needs to be done so we don't end up reinventing the wheel.

I'll be happy to clarify/assist if you have any doubts/questions, feel free to PM me (or just comment here).

**Comment by immortanjoe** [comment]
I also have the same question. If webrtc is used, the signal server and relay server can be reused.

**Comment by VarLad** [comment]
I believe this would also make it possible to connect to other PCs P2P without being on the same network. 
This would mean a lot less load for RustDesk's servers as they'll only be used when P2P connection fails (for example, for VPN users), and for making the P2P connection.
@rustdesk Any idea if there's any plans to work on this? This seems like a very good idea and I've seen other projects (like https://github.com/saljam/webwormhole ) establish direct P2P connection between users using WebRTC as well.

**Comment by zuoxinyu** [comment]
As for I know, [chrome remote desktop](https://remotedesktop.google.com/support/) uses WebRTC as the core protocol. And I'm working on a demo RDP app based on WebRTC native C++ SDK.  

It's easy to implement basic things like media/file/input events transportation and P2P connection. Yes, it just works, as said `Basically almost as simple as making an HTTP request` . 

But it's not easy to achieve chrome level optimization. Chrome remote desktop does a bunch of work on dealing with codecs/signlaling/congest control/and more, which are highly depended on chromium codes. 

Besides, the webrtc sdk is extremely undocumented and rapidly iterating, anyway, WebRTC is for **WEB** , not for native .

**Reply by WofWca** [reply]
I mean, what are the alternatives. I propose WebRTC mainly as a transport thing, and it very much can transport arbitrary data very easily and flexibly. The media part is a bonus, although you say it is not good in the SDK.

**Comment by shi-yan** [comment]
We used to work on cloud gaming/ remote desktop. WebRTC hasn't been a great choice, (hence google added webtransport and webcodec).

WebRTC was designed for video chat, not streaming desktop. They have different requirements regarding buffering, quality and latency, etc. And like most web standards, it is shipped as a blackbox without much tweak-ability .

**Reply by WofWca** [reply]
Thanks for the heads-up, but again, you're talking about the built-in media support. And I'm proposing WebRTC as a transport thing, which can be used to transport arbitrary data.

**Comment by TCB13** [comment]
What is instead of WebRTC that wasn't built for this we simply have UPNP/NAT-PMP support like **any other solution**? https://github.com/rustdesk/rustdesk/discussions/2730

I don't get why UPNP wasn't implemented from day zero. This allows clients to connect directly to each other and avoid overloading the servers. It is kind of crazy not to use it.

**Comment by blackfireball5** [comment]
Has @rustdesk chimed in on this at all? Seems WebRTC would make so much sense for Rustdesk. It's been a few years since this discussion was started. Curious if there are any plans in the works.

**Reply by rustdesk** [reply]
Sorry, no. This is open source, you do not need to wait for us, you can do it yourself.

**Reply by rustdesk** [reply]
Lock this discussion in case someone @ me again.
